pkg/reconciler/pipelinerun/tracing.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2022 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pipelinerun
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/trace"
+	"knative.dev/pkg/logging"
+)
+
+const (
+	// TracerName is the name of the tracer
+	TracerName = "PipelineRunReconciler"
+	// SpanContextAnnotation is the name of the Annotation for storing SpanContext
+	SpanContextAnnotation = "tekton.dev/pipelinerunSpanContext"
+	// TaskRunSpanContextAnnotation is the name of the Annotation used for propogating SpanContext to TaskRun
+	TaskRunSpanContextAnnotation = "tekton.dev/taskrunSpanContext"
+)
+
+// initialize tracing by creating the root span and injecting the
+// spanContext is propogated through annotations in the CR
+func initTracing(ctx context.Context, tracerProvider trace.TracerProvider, pr *v1beta1.PipelineRun) context.Context {
+	logger := logging.FromContext(ctx)
+	pro := otel.GetTextMapPropagator()
+
+	// SpanContext was created already
+	if pr.Status.SpanContext != nil && len(pr.Status.SpanContext) > 0 {
+		return pro.Extract(ctx, propagation.MapCarrier(pr.Status.SpanContext))
+	}
+
+	spanContext := make(map[string]string)
+
+	// SpanContext was propogated through annotations
+	if pr.Annotations != nil && pr.Annotations[SpanContextAnnotation] != "" {
```

Comment:
Nice! These means that a `PipelineRun` too can receive an input context, which means we can support existing contexts for pipelines, like in the case of pipeline in pipeline, or a pipeline triggered via CloudEvent with a tracing context in it.